src/uct/ib/base/ib_iface.c

```diff
@@ -821,11 +828,12 @@ static ucs_status_t uct_ib_iface_init_pkey(uct_ib_iface_t *iface,
         if ((config->pkey == UCS_HEXUNITS_AUTO) ||
             /* take only the lower 15 bits for the comparison */
             ((pkey & UCT_IB_PKEY_PARTITION_MASK) == config->pkey)) {
-            if (!(pkey & UCT_IB_PKEY_MEMBERSHIP_MASK) &&
-                /* limited PKEY has not yet been found */
-                (lim_pkey == UCT_IB_ADDRESS_INVALID_PKEY)) {
-                lim_pkey_index = pkey_index;
-                lim_pkey       = pkey;
+            if (!(pkey & UCT_IB_PKEY_MEMBERSHIP_MASK)) {
+                if (lim_pkey == UCT_IB_ADDRESS_INVALID_PKEY) {
```

Comment:
why nested if?

Reply:
it was a bug, we set new found limited PKEY if limited PKEY was already set.

Reply:
i would rewrite the code like this:
```
if (pkey & UCT_IB_PKEY_MEMBERSHIP_MASK) {
   iface->pkey_index = pkey_index;
   iface->pkey       = pkey;
   pkey_found        = 1;
} else if (lim_pkey == UCT_IB_ADDRESS_INVALID_PKEY) {
    lim_pkey_index = pkey_index;
    lim_pkey = pkey;
}
```

Reply:
good idea, done